doc/source/OEP/OEP_002.rst

```diff
@@ -0,0 +1,239 @@
+===================
+ OEP 2 : read-mode
+===================
+
+
+.. contents::
+   :local:
+
+Status
+======
+
+**Discussion**
+
+
+Branches and Pull requests
+==========================
+
+
+Abstract
+========
+
+This OEP proposes to add a 'read mode' to ophyd devices.  Ophyd
+devices may be read in different modes that we want to access with in
+the same 'stage' context.  For example, one could imagine a streaming
+mode where an ophyd object upon calling its
+``obj.read(read_mode='streaming')`` immediately after triggering a
+detector to get an event which can be consumed 'live' and then later
+calling ``obj.read()`` after the detector has finished.  The two reads
+may contain different keys (as not all children may support streaming).
+
+A second example would be a detector which requires dark frames to be
+taken and stored.  This could be implemented via
+``obj.read(read_mode='dark')`` which would return the dark field. and
+``obj.read()`` would return the bright field.
+
+Another example is a detector with a scalar and array mode where
+``obj.read(read_mode='scalar')`` would just grab the current value and
+``obj.read()`` would read everything.
+
+
+Detailed description
+====================
+
+Changes to function signatures
+------------------------------
+
+To support this change we will have to add an optional keyword
+argument ``read_mode`` to ``trigger``, ``read``, and ``describe``.  If
+a device accepts the kwarg (and a particular value of the kwarg) on
+any one of the three methods then it must accept it on all of them.
+The values accepted for ``read_mode`` should be strings.
+
+Objects much have a 'default' read mode so ``read()`` works.
+
+``trigger`` may dis-regard ``read_mode``, but must accept it if
+``read`` does.
+
+An object may require a matched ``trigger(read_mode=FOO)`` to go with
+a ``read(read_mode==FOO)``.
+
+Changes to the RunEngine
+------------------------
+
+The descriptor cache will now need to be by keyed by ``(obj,
+msg.kwargs.get('read_mode', None))`` rather than just ``obj`` and the
+descriptor building logic will have to be updated to account for both
+the object and read mode involved.  Reading the same device twice with
```

Comment:
"the descriptor building logic will have to be updated to account for both the object and read mode involved"
About implementing that. Do we want bluesky to know about `read_mode` or just pass kwargs (should we explicitly pass `read_mode` into the descriptor)? I would be tempted to say we should pass `read_kwargs` to the descriptor. What do you think?

Reply:
I can see arguments either way.  Currently the `_read` coro passes all the args and kwargs into `obj.read` in case we need to, on the fly, write a deeply coupled custom device and plan we can.

If we put _all_ of the args/kwargs in the hash key we may be cutting off some flexibility (ex, if you have weird device that needs some kwarg passed to it that is different every time).  If every key went in then you would end up with a different descriptor every time which is not what we want.  But, the next time we want to add a 'standard' kwarg to 'read' and friends we would not have to change any of the descriptor logic.

If we put only read_mode we keep the flexibility to do really ad-hoc custom stuff if needed, but the next time we want to add a new 'standard' kwarg we will have to make bigger changes.

I was primarily thinking of the logic required to determine if for the set of objects read in this event if we need to construct a new descriptor or if we have a cached one that works or we need to fail, but these things should also be put into the descriptor document.  Any suggestions as to where?

I lean towards making 'read_mode' special to preserve the 'do something really wacky' flexibility.

Reply:
That's a good point and something we should discuss first: how to know when a generated descriptor is stale.

As far as I understand, this is done by sending the `configure` message. Since the logic supplying the `kwargs` for read is at the same level as the logic for flagging a descriptor as stale, is this enough? I would be more tempted to say that such flagging should remain on the bluesky side, and the run engine shouldn't handle this.